feat(virtctl): Change separators in virtctl port-forward syntax

0xFelix opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does

This changes the target syntax of virtctl port-forward/ssh/scp to
'type/name[/namespace]'. This allows to properly support resources with
dots in their names while keeping extended semantics for compatibility
with SSH config files.

SCP does not support slashes in hostnames, therefore the target
generated for wrapped SSH/SCP clients is changed to
$TYPE.$NAME.$NAMESPACE, to keep support with SSH/SCP, while also
providing a unique name for known hosts files. This is a non-breaking
change for users of virtctl ssh/scp.

To keep compatibility with existing scripts for now, legacy parsing
of the syntax type/name.namespace is activated when no namespace is
specified in the target.

Normalize the type when parsing a target the was passed to virtctl
port-forward/ssh/scp to the singular abbreviated variant of the
type.

Before this PR:

virtctl port-forward/ssh/scp has syntax type/name[.namespace]

After this PR:

virtctl port-forward/ssh/scp has syntax type/name[/namespace]
and supports resources with dots in their name properly now.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

Merge after https://github.com/kubevirt/kubevirt/pull/13938

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
The virtctl port-forward/ssh/scp syntax was changed to type/name[/namespace]. It now supports resources with dots in their name properly.
```

0xFelix commented:
/retest

fossedihelm commented on pkg/virtctl/portforward/portforward.go:
```diff
@@ -137,12 +137,12 @@ func (o *PortForward) prepareCommand(args []string, fallbackNamespace string) (k
 
 func (o *PortForward) setResource(kind, namespace string, client kubecli.KubevirtClient) error {
-	if kindIsVMI(kind) {
+	if kind == "vmi" {
```
IMHO "vmi"|"vm" should be constants now

fossedihelm reviewed (commented):
Thank you @0xFelix
It looks good overall; just a few comments below.
Also, did you consider using a regex for type/name[/namespace] parsing?

0xFelix commented:
@fossedihelm Addressed your comments. The idea of a regex is nice, but to me that makes the code harder to maintain in the future and with a regex it is also not possible to return specific errors. I chose to stay with manual parsing therefore.

fossedihelm reviewed (commented):
/approve

akrejcir commented:
I didn't review the code yet, but did you consider using the format `type/[namespace/]name` ? In my opinion it would be consistent with how k8s uses URL for API requests.

jcanocan commented on pkg/virtctl/portforward/portforward.go:
```diff
@@ -212,66 +219,68 @@ func examples() string {
-			return "", "", "", errors.New("unsupported resource kind " + kind)
+		name = parts[1]
+		namespace = parts[2]
+		if namespace == "" {
+			return "", "", "", errors.New("namespace cannot be empty")
```
Nit: I would extract this check outside the switch case, as we do with `name` and `kind`.

0xFelix replied:
It applies only in this specific case. In others namespace is allowed to be empty.

jcanocan replied:
I understand now. Thanks for the clarification.

jcanocan commented on pkg/virtctl/portforward/portforward.go:
```diff
@@ -212,66 +219,68 @@ func examples() string {
-		kind == "vms" ||
-		kind == "virtualmachine" ||
-		kind == "virtualmachines"
+func normalizeKind(kind string) (string, error) {
+	switch kind {
+	case vm, "vms", "virtualmachine", "virtualmachines":
+		return vm, nil
+	case vmi, "vmis", "virtualmachineinstance", "virtualmachineinstances":
+		return vmi, nil
+	}
+	return "", fmt.Errorf("unsupported resource type '%s'", kind)
 }
```
Would be a good idea to lower case the `kind` string before the switch case?

0xFelix replied:
Good idea! Updated

jcanocan reviewed (commented):
Good job @0xFelix!
Just a couple of silly nits.

ksimon1 commented on pkg/virtctl/portforward/portforward_test.go:
```diff
@@ -19,20 +19,36 @@ var _ = Describe("Port forward", func() {
+		Entry("kind vm", "vm/testvm", "", "testvm", "vm", ""),
+		Entry("kind vms", "vms/testvm", "", "testvm", "vm", ""),
+		Entry("kind virtualmachine", "virtualmachine/testvm", "", "testvm", "vm", ""),
+		Entry("kind virtualmachines", "virtualmachines/testvm", "", "testvm", "vm", ""),
```
Would you please add one more test case with incorrect kind?

ksimon1 commented:
/lgtm

0xFelix commented:
> I didn't review the code yet, but have you considered using the format type/[namespace/]name ? In my opinion it would be consistent with how k8s uses URL for API requests.

@akrejcir The idea was to build upon what `kubectl port-forward` does and extend it with a namespace. If the namespace is omitted, it works exactly like `kubectl port-forward`.

```
$ kubectl port-forward -h
[...]
Usage:
  kubectl port-forward TYPE/NAME [options] [LOCAL_PORT:]REMOTE_PORT [...[LOCAL_PORT_N:]REMOTE_PORT_N]
```

0xFelix commented:
/retest

codingben commented on pkg/virtctl/scp/scp_test.go:
```diff
@@ -16,15 +16,15 @@ var _ = Describe("SCP", func() {
 		),
 		Entry("copy from remote location",
-			"cirros@vmi/remote.mynamespace:myfile.yaml", "myfile.yaml",
+			"cirros@vmi/remote/mynamespace:myfile.yaml", "myfile.yaml",
```
Is not usually `namespace/object`? We don't follow this k8s convention here?

codingben reviewed (approved):
/lgtm

dbasunag commented:
This would impact upgrade lane tests that uses `virtctl port-forward` to ensure vm that was created before upgrade, stays reachable post upgrade. Any direct upgrade path to 4.19 would be impacted, unless we have both `.` and `/` work for a release.
